### DISCUSSION TITLE
Is it possible to call the loader function passing parameters that are not in the URL?

### DISCUSSION THREAD

**Discussion by FranciscoGileno** [remix-run/react-router#9964] [Q&A]
I have a loader function that calls the apollo client to fetch the data:

```js
export async function loader({ request }) {
  const url = new URL(request.url)
  const variables = parseSearchParams(url.search)
  const queryResponse = await client.query(myQuery, { variables })
  return json(queryResponse.data)
}
```
1. My variables are built from the URL search params.
2. The `client.query` caches the result, so if the same search params are sent, it gets from the cache.

For 99% of the time, it works great.

But when I click on the Refresh button, I don't want to use the cache for the same search params.

If I call `navigate` passing the search Param: `refresh=true`, I can get this parameter in the loader and query from server. However, this parameter will appear in the URL, which I prefer not to.

Is it possible to call the loader function passing parameters that are not in the URL?

**Comment by bodokaiser** [comment]
@FranciscoGileno How did you structure your app to have apollo `client` available in the `loader`?

**Comment by FranciscoGileno** [comment]
Hi @bodokaiser .

That's just a export / import.. 

On lib/apollo
```js
export const client = new ApolloClient({
  link: ApolloLink.from([redirectLink, commonLink]),
  cache,
  defaultOptions
})
```

On Loader:

```ts
import { client } from 'lib/apollo'
...
export async function loader({ request }) {
  const url = new URL(request.url)
  const variables = parseSearchParams(url.search)
  const queryResponse = await client.query(myQuery, { variables })
  return json(queryResponse.data)
}
```

**Reply by bodokaiser** [reply]
Hi @FranciscoGileno,

What kind of authentication do you use with Apollo?

We are using AWS Amplify (Cognito) which requires a call to `Auth.configure({ ... })` at module runtime before you can get the jwt token via `await Auth.currentSession()`.
As `createBrowserRouter()` is also called at module runtime, I run into an error of the `Auth` module not being configured yet.

In principle, one could use a dependency injection pattern alike

```tsx
export function createRouter(client) {
  return createBrowserRouter([
    //
  ])
}
```

but this also forces us to define any route and loaders inside that closure.

But maybe I am missing something here?
